webapp/store/models.py

```diff
@@ -0,0 +1,172 @@
+import markdown
+import gfm
+import re
+
+
+cs = CharmStore("https://api.jujucharms.com/v5")
+
+def get_charm_or_bundle(name):
+    try:
+        entity_data = cs.entity(name, True)
+    except Exception:
+        return abort(404, "Entity not found {}".format(name))
+
+    charm_or_bundle_id = entity_data['Id']
+    ref = references.Reference.from_string(charm_or_bundle_id)
+    meta = entity_data.get('Meta', None)
+    bzr_url, revisions = _extract_from_extrainfo(meta, ref)
+    bugs_url, homepage = _extract_from_commoninfo(meta)
+    revision_list = meta['revision-info'].get('Revisions')
+
+    charm_or_bundle = {
+        'archive_url': cs.archive_url(ref),
+        'bugs_url': bugs_url,
+        'bzr_url': bzr_url,
+        'entity_data': entity_data,
+        'files': _getEntityFiles(ref, meta['manifest']),
+        'homepage': homepage,
+        'icon': _parse_charm_icon(charm_or_bundle_id),
+        'id': charm_or_bundle_id,
+        'latest_revision': _expand_revision(revision_list[0], meta['charm-metadata']['Name']),
+        'options': meta.get('charm-config', {}).get('Options'),
+        'owner': entity_data.get('owner', {}).get('User'),
+        'provides': meta['charm-metadata'].get('Provides'),
+        'requires': meta['charm-metadata'].get('Requires'),
+        'resources': _extract_resources(ref, meta.get('resources')),
+        'readme': _render_markdown(cs.entity_readme_content(charm_or_bundle_id)),
+        'revision_list': revision_list,
+        'revision_number': ref.revision,
+        'revisions': revisions,
+        'series': meta.get('supported-series', {}).get('SupportedSeries'),
+        'is_charm': 'charm-metadata' in meta
+    }
+
+    return charm_or_bundle
+
+
+def _getEntityFiles(ref, manifest=None):
+    try:
+        files = cs.files(ref, manifest=manifest)
+    except EntityNotFound:
+        files = {}
+    if not files:
+        files = {}
+    try:
+        readme = cs.entity_readme_content(ref)
+    except EntityNotFound:
+        readme = None
+    return {
+        'files': files,
+        'readme': readme,
+    }
+
+
+def _parse_charm_icon(charm_id):
+    try:
+        return cs.charm_icon_url(charm_id)
+    except Exception:
+        return abort(500, "Entity not found {}".format(charm_id))
+
+
+def _expand_revision(revision_id, charm_name):
+    '''From a revision id, create a Dict.
+    @param revision_id A revision id string.
+    @param charm_name The name of the charm
+    '''
+
+    id = revision_id.split('-')[-1]
+    url = charm_name + '/' + id
+    return {
+        'id': int(id),
+        'full_id': revision_id,
+        'url': url}
+
+
+def _extract_resources(ref, data):
+    """Extract data from resources metadata.
+    @param ref The reference of the entity.
+    @param data the resources metadata associated with the entity.
+    @return a dictionary of resource name and an array containing
+            the file extension, the file link of the resource.
+    """
+    if not data:
```

Comment:
I'm not sure this is necessary. Are we expecting to hit this block when `data == {}`? Or when `data == None`? Because if `data == {}` then we could happily continue, and `for i in data:` would simply not do anything, and we'd ultimately return `{}` anyway.

The only usage I can see of this function is line 41:

``` python
'resources': _extract_resources(ref, meta.get('resources')),
```

If you replaced it with:

``` python
'resources': _extract_resources(ref, meta.get('resources', {})),
```

Then you definitely wouldn't need this block.